cpp/src/quantiles/tdigest/tdigest_aggregation.cu

```diff
@@ -983,47 +993,95 @@ struct typed_reduce_tdigest {
-  size_type const* inner_offsets;
+struct group_num_clusters_func {
+  GroupOffsetsIter group_offsets;
+  size_type const* tdigest_offsets;
 
   __device__ size_type operator()(size_type group_index)
   {
-    auto const tdigest_begin = outer_offsets[group_index];
-    auto const tdigest_end   = outer_offsets[group_index + 1];
-    return inner_offsets[tdigest_end] - inner_offsets[tdigest_begin];
+    auto const tdigest_begin = group_offsets[group_index];
+    auto const tdigest_end   = group_offsets[group_index + 1];
+    return tdigest_offsets[tdigest_end] - tdigest_offsets[tdigest_begin];
   }
 };
 
-// utility for merge_tdigests.
+/**
+ * @brief Function to determine if a group is empty.
+ *
+ * Used in `merge_tdigests`.
+ */
 struct group_is_empty {
   __device__ bool operator()(size_type group_size) { return group_size == 0; }
 };
 
-// utility for merge_tdigests.
+/**
+ * @brief Functor that returns the grouping key for each tdigest cluster.
+ *
+ * Used in `merge_tdigests`.
+ */
 template <typename GroupLabelsIter>
 struct group_key_func {
   GroupLabelsIter group_labels;
-  size_type const* inner_offsets;
-  size_type num_inner_offsets;
+  size_type const* tdigest_offsets;
+  size_type num_tdigest_offsets;
 
+  /**
+   * @brief Returns the group index for an absolute cluster index. The index `n` is the index of the
+   * `n`-th non-empty cluster.
+   */
   __device__ size_type operator()(size_type index)
   {
     // what -original- tdigest index this absolute index corresponds to
-    auto const iter = thrust::prev(
-      thrust::upper_bound(thrust::seq, inner_offsets, inner_offsets + num_inner_offsets, index));
-    auto const tdigest_index = thrust::distance(inner_offsets, iter);
+    auto const iter          = thrust::prev(thrust::upper_bound(
+      thrust::seq, tdigest_offsets, tdigest_offsets + num_tdigest_offsets, index));
+    auto const tdigest_index = thrust::distance(tdigest_offsets, iter);
 
     // what group index the original tdigest belongs to
     return group_labels[tdigest_index];
   }
 };
 
+/**
+ * @brief Perform a merge aggregation of tdigests. This function usually takes the input as the
+ * outputs of multiple `typed_group_tdigest` calls, and merges them.
+ *
+ * A tdigest cluster can be empty in the input, which means that there was no valid input data to
+ * generate that cluster. These empty clusters are currently stored differently in different parts
+ * of the tdigest column. They are "compressed" in the `means`, `weights`, and `offsets` columns,
+ * but not in the `min` and `max` columns.
+ *   - The `means` and `weights` columns do not contain values for empty clusters.
+ *   - The empty clusters are stored as two consecutive same values. For example, given an offsets
+ * column of (0, 1, 1, 2), the second cluster where its offset is 1 is empty. Note that the offsets
+ * are the offsets for the means and the weights.
+ *   - The `min` and `max` columns contain 0s for the empty clusters.
```

Comment:
This comment may be a bit unnecessary/inaccurate.  Empty lists of things stemming from two consecutive offsets is a very common thing throughout cudf.  For example, we can have an empty row in a string column where there are no corresponding chars, but we still have an entry in the validity bitmask.

So in this case, it's not really 'compressed'.  There is a value in the size of this cluster:  offsets[cluster+1] - offsets[cluster].  It just happens to be zero.

Reply:
> So in this case, it's not really 'compressed'. There is a value in the size of this cluster: offsets[cluster+1] - offsets[cluster]. It just happens to be zero.

I used the term "compressed" in a more generic sense that the actual values, such as nulls, are not stored for empty clusters. Yes, we store the offsets for empty clusters, but not the means and weights. The "compressed" is commonly used in this way in data compressions. Data is compressed by not storing some values explicitly. That being said, I can rephrase this doc if it is inconsistent with what it usually means in libcudf and thus may cause confusion.

> Empty lists of things stemming from two consecutive offsets is a very common thing throughout cudf. For example, we can have an empty row in a string column where there are no corresponding chars, but we still have an entry in the validity bitmask.

I'm not sure if this doc is unnecessary. What I wanted to point out here is that empty clusters are handled differently in the mins and maxes, and the weights and means. Someone could argue that, because the mins and maxes are the sub-columns that we don't store offsets for them, it is natural to have some placeholders for empty clusters in them. But they are the sub-columns of the tdigest column as well as the centroids sub-column in which empty clusters are implicit. The inconsistency between them seems worth to me to point out in the doc.

Reply:
```
 * A tdigest cluster can be empty in the input, which means that there was no valid input data to
 * generate that cluster. These empty clusters are currently stored differently in different parts
 * of the tdigest column. They are more explicit in the `min` and `max` columns than in the `means`,
 * `weights` columns.
 *   - The `means` and `weights` columns do not contain values for empty clusters.
 *     - In the `offsets` column for the means and weights, the offsets to empty clusters are
 *      stored so that their size is 0. For example, given an offsets column of (0, 1, 1, 2),
 *     the second cluster where its both start and end offsets are 1 is empty.
 *   - The `min` and `max` columns contain 0s for the empty clusters.
```

I have rephrased it little. Do you think it makes more sense now?

Reply:
I think this is still confusion about the (Arrow) structure of the tdigest column. The column is structured as:
```
 * struct {
 *   // centroids for the digest
 *   list {
 *    struct {
 *      double    // mean
 *      double    // weight
 *    }
 *   }
 *   // these are from the input stream, not the centroids. they are used
 *   // during the percentile_approx computation near the beginning or
 *   // end of the quantiles
 *   double       // min
 *   double       // max
 * }
```

Each row in the column is 1 tdigest, which has a min, max and a list of centroids (which may be empty). As with any list column, the contents of a row may be empty, but there will always be one offset per row.  So the description is just reiterating "this is how a cudf column works".  I think it might be easier to say:

```
* A tdigest cluster can be empty in the input, which means that there was no valid input data to
 * generate that cluster. These empty tdigests will have no centroids (means or weights) and will
 * have a `min` and `max` of 0.
```

Reply:
Also, "clusters" is being misused here a bit.  There's no concept of clusters within a tdigest.  The clusters we build are part of the actual generation (or merge) process - figuring out what centroids in the input are close enough that they can be clustered together and ultimately squeeze together.

Reply:
Sounds good. Updated as suggested.